packages/eas-cli/src/build/runBuildAndSubmit.ts

```diff
@@ -587,3 +608,56 @@ async function validateExpoUpdatesInstalledAsProjectDependencyAsync({
     }
   }
 }
+
+function isEnvironment(env: string): env is EnvironmentVariableEnvironment {
+  if (
+    Object.values(EnvironmentVariableEnvironment).includes(env as EnvironmentVariableEnvironment)
+  ) {
+    return true;
+  }
+  return false;
+}
+
+async function resolveEnvVarsAsync({
+  sessionManager,
+  flags,
+  buildProfile,
+  graphqlClient,
+  projectId,
+}: {
+  sessionManager: SessionManager | undefined;
+  flags: BuildFlags;
+  buildProfile: ProfileData<'build'>;
+  graphqlClient: ExpoGraphqlClient;
+  projectId: string;
+}): Promise<Record<string, string>> {
+  const environment = flags.environment ?? buildProfile.profile.environment;
+
+  if (!environment || !isEnvironment(environment)) {
+    return {};
+  }
+  const profileEnv = buildProfile.profile.env || {};
+
+  if (sessionManager) {
+    try {
+      const environmentVariables = await withSudoModeAsync(
+        sessionManager,
+        async () =>
+          await EnvironmentVariablesQuery.byAppIdWithSensitiveAsync(graphqlClient, {
+            appId: projectId,
+            environment,
+          })
+      );
+      const envVars = Object.fromEntries(
+        environmentVariables
+          .filter(({ name, value }) => name && value)
+          .map(({ name, value }) => [name, value])
+      ) as Record<string, string>;
+      return { ...profileEnv, ...envVars };
```

Comment:
I think we should be verbose here about what env vars we are loading and from where, this will help users to understand their setup better

We should probably replace https://github.com/expo/eas-cli/blob/75b310e7e08ccf5c35cc224402a9e78385e04c8e/packages/eas-cli/src/build/runBuildAndSubmit.ts#L132-L138 with new messages